runtime/ISO_Fortran_binding.cc

```diff
@@ -193,20 +199,150 @@ int CFI_is_contiguous(const CFI_cdesc_t *descriptor) {
   return 1;
 }
 
+static inline bool IsAssumedSize(const CFI_cdesc_t *dv) {
+  return (dv->rank > 0 && dv->dim[dv->rank - 1].extent == -1);
+}
+
 int CFI_section(CFI_cdesc_t *result, const CFI_cdesc_t *source,
     const CFI_index_t lower_bounds[], const CFI_index_t upper_bounds[],
     const CFI_index_t strides[]) {
-  return CFI_INVALID_DESCRIPTOR;  // TODO
+  CFI_index_t extent[CFI_MAX_RANK];
+  CFI_index_t actualStride[CFI_MAX_RANK];
+  CFI_rank_t resRank{0};
+  char *shiftedBaseAddr{reinterpret_cast<char *>(source->base_addr)};
+
+  if (source->rank == 0) {
+    return CFI_INVALID_RANK;
+  }
+  if (IsAssumedSize(source) && upper_bounds == nullptr) {
+    return CFI_INVALID_DESCRIPTOR;
+  }
+  if ((result->type != source->type) ||
+      (result->elem_len != source->elem_len)) {
+    return CFI_INVALID_DESCRIPTOR;
+  }
+  if (result->attribute == CFI_attribute_allocatable) {
+    return CFI_INVALID_ATTRIBUTE;
+  }
+  if (source->base_addr == nullptr) {
+    return CFI_ERROR_BASE_ADDR_NULL;
+  }
+
+  bool isZeroSized{false};
+  for (int j{0}; j < source->rank; ++j) {
+    const CFI_dim_t &dim = source->dim[j];
+    const CFI_index_t srcLB = dim.lower_bound;
+    const CFI_index_t srcUB = srcLB + dim.extent - 1;
+    const CFI_index_t lb = (lower_bounds != nullptr) ? lower_bounds[j] : srcLB;
+    const CFI_index_t ub = (upper_bounds != nullptr) ? upper_bounds[j] : srcUB;
+    const CFI_index_t stride = (strides != nullptr) ? strides[j] : 1;
+
+    if (stride == 0 && lb != ub) {
+      return CFI_ERROR_OUT_OF_BOUNDS;
+    }
+    if ((lb <= ub && stride >= 0) || (lb >= ub && stride < 0)) {
+      if ((lb < srcLB) || (lb > srcUB) || (ub < srcLB) || (ub > srcUB)) {
+        return CFI_ERROR_OUT_OF_BOUNDS;
+      }
+      shiftedBaseAddr += (lb - srcLB) * dim.sm;
+      extent[j] = (stride != 0) ? 1 + (ub - lb) / stride : 1;
+    } else {
+      isZeroSized = true;
+      extent[j] = 0;
+    }
+    actualStride[j] = stride;
+    resRank += (stride != 0);
+  }
+  if (resRank != result->rank) {
+    return CFI_INVALID_DESCRIPTOR;
+  }
+
+  // For zero-sized arrays, base_addr is proc-dep (see 18.5.3).
+  // We keep it on the source base_addr
+  result->base_addr = isZeroSized ? source->base_addr
+                                  : reinterpret_cast<void *>(shiftedBaseAddr);
```

Comment:
Here the cast isn't needed at all. But if you think it makes it clearer, use static_cast.

Reply:
No, I do not think it adds a lot, I have removed it that makes it more readable.  Thanks